### PR TITLE
Jetpack Cloud (Agencies): Refactor license issue/assign hook

### PR DESCRIPTION
**Pull request by mattgawarecki**
Resolves `1202619025189113-as-1203377005312989`.
Related to #70101.

## Proposed Changes

* Move the hook for issuing and assigning licenses into its own file.
* Extract several auxiliary functions and hooks from the main hook logic, to separate concerns and improve readability.
* Make the list of product slugs to issue an argument to the hook's callback, rather than a parameter of the hook itself.
* Rename the main hook to `useIssueAndAssignLicenses`, to more accurately describe what it does (i.e., issue and assign product licenses).
* Correct misspellings in a few translated strings and Tracks event names.
* Add/update inline documentation.

## Testing Instructions

_**TL;DR:** Full regression testing of the "issue license(s)" flow in the dashboard._

Run through the following scenarios, verifying that license(s) are issued and assigned correctly, with no unwanted "extra" licenses created, no errors with licenses that you request, and appropriate page redirects when the occasion arises.

* Provision one individual (non-bundle) product license from the dashboard. Pay special attention to the success notification for misspellings and proper bolding of the product name.
* Provision two individual product licenses. Pay special attention to the success notification for proper grammar and bolding of product names.
* Provision three individual product licenses. Pay special attention to the success notification for proper grammar and bolding of product names.
* Provision a bundle. Pay special attention to the list of provisioned licenses afterward and ensure only one license was issued/assigned.

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - 0-as-1203377005312989

### CODE REVIEW

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/8068331

Hi @mattgawarecki, could you please edit the description of this PR and add a screenshot for our translators? Ideally it'd include all of the following strings:
- `{{strong}}%(first)s{{/strong}} and {{strong}}%(second)s{{/strong}} were successfully issued.`
- `{{strong}}%(initialProducts)s{{/strong}}, and {{strong}}%(finalProduct)s{{/strong}} were successfully issued.`

Thank you in advance!

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/8068331

Hi @mattgawarecki, could you please edit the description of this PR and add a screenshot for our translators? Ideally it'd include all of the following strings:
- `{{strong}}%(first)s{{/strong}} and {{strong}}%(second)s{{/strong}} were successfully issued.`
- `{{strong}}%(initialProducts)s{{/strong}}, and {{strong}}%(finalProduct)s{{/strong}} were successfully issued.`
- `WooCommerce Extensions:`

Thank you in advance!

File: client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+			const issuedKeys = issuedLicenses.map( ( { value } ) => value.license_key );
+			const issuedProducts = issuedKeys
+				.map( ( key ) => {
+					const productSlug = key.split( '_' )[ 0 ];
+					return products?.data?.find( ( p ) => p.slug === productSlug );
+				} )
+				.filter( ( p ): p is APIProductFamilyProduct => p !== undefined );
+
+			// Exit early if we don't see any issued licenses matching a product we know
+			if ( issuedProducts.length === 0 ) {
```
Comment (jkguidaven) [inline client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx]: I'm curious when exactly we have a scenario like this.  Is this something that happens when we provide an invalid product slug?

File: client/jetpack-cloud/sections/partner-portal/issue-multiple-licenses-form/index.tsx [hunk 2]
```diff
@@ -127,103 +124,108 @@ export default function IssueMultipleLicensesForm( {
...
+				recordTracksEvent(
+					'calypso_partner_portal_issue_bundle_license_with_existing_standalone_products'
+				)
+			);
+		}
+
+		issueAndAssignLicenses( [ selectedBundle ] );
+	}, [ dispatch, hasPurchasedProductsWithoutBundle, issueAndAssignLicenses, selectedBundle ] );
+
+	if ( isLoadingProducts ) {
```
Comment (jkguidaven) [inline client/jetpack-cloud/sections/partner-portal/issue-multiple-licenses-form/index.tsx:153]: 👍

File: client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+
+	return useMemo( () => {
+		const isReady = issueLicenses.isReady && assignLicensesToSite.isReady;
+
+		const issueAndAssignLicenses = async ( selectedProducts: string[] ) => {
+			if ( ! isReady || selectedProducts.length === 0 ) {
+				return;
+			}
+
+			dispatch(
```
Comment (jkguidaven) [inline client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx]: Not a blocker and just nitpicking.

Not sure we can use `recordTracksEvent` from this package here but that one is more leaner since we don't need to use `dispatch`.

`
import { recordTracksEvent } from '@automattic/calypso-analytics';
`

**Review by jkguidaven — APPROVED** [review]
@mattgawarecki  Thanks for significantly improving the readability of the code. :+1: 

I did quick testing on the PR, and it is working as expected. Pages are redirecting correctly, and licenses are issued and assigned correctly.

While I left a comment but is non-blocker and just nitpicking. 

I also suggest having @Automattic/jetpack-avalon  look at this PR to be sure we didn't miss anything.

File: client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx (hunk 1 above)
Comment (mattgawarecki) [reply client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx]: Honestly? I'm not sure. But since the main goal of this PR was to refactor, I tried to preserve all the existing behavior I could, whether or not I could determine if it made 100% sense.

File: client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx (hunk 3 above)
Comment (mattgawarecki) [reply client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx]: According to the [documentation in the `calypso-analytics` package](https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-analytics/README.md):

> In most situations it is best to use the [Analytics Middleware](https://github.com/Automattic/wp-calypso/tree/HEAD/client/state/analytics), which has no direct browser dependencies and therefore will not complicate any unit testing of the modules where it is used.

File: client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+			);
+
+			return Promise.allSettled( requests );
+		};
+
+		return {
+			isReady: issueLicenseMutation.isIdle,
+			issueLicenses,
+		};
+	}, [ issueLicenseMutation ] );
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx]: Hooks tend to return fresh objects every time (which defeats the memoization) so it's usually better to depend on the particular properties needed i.e.`[issueLicenseMutation.mutateAsync, issueLicenseMutation.isIdle]`.
A very terse check of react-query's source **seems** to confirm this:
https://github.com/TanStack/query/blob/b18426da86e2b8990e8f4e7398baaf041f77ad19/packages/react-query/src/useMutation.ts#L122

File: client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+			);
+
+			return Promise.allSettled( requests );
+		};
+
+		return {
+			isReady: assignLicenseMutation.isIdle,
+			assignLicensesToSite,
+		};
+	}, [ siteId, assignLicenseMutation ] );
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx]: See the other comment about hook return values and memo dependencies.

File: client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+			}
+
+			dispatch(
+				recordTracksEvent( 'calypso_partner_portal_multiple_licenses_issued', {
+					products: issuedProducts.join( ',' ),
+				} )
+			);
+
+			// If no site is selected, announce that licenses were issued;
+			// then, redirect to somewher more appropriate
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx]:
```suggestion
			// then, redirect to somewhere more appropriate
```

File: client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+			}
+
+			// If a specific site is already selected,
+			// let's assign the licenses we just issued to it
+			const assignLicenseResponses = await assignLicensesToSite.assignLicensesToSite( issuedKeys );
+
+			const assignedLicenses = assignLicenseResponses
+				.filter( ( p ): p is PromiseFulfilledResult< APILicense > => p.status === 'fulfilled' )
+				.map( ( { status, value } ) => {
+					const productSlug = value.license_key.split( '_' )[ 0 ];
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx]: A while ago I realized that siteless licenses actually can include multiple underscores in the license key so here we should be splitting on the last underscore instead of the first.
Also, if we don't have something in place already, we should move this to a reusable function so we don't have magic license key string manipulation logic in multiple places.

File: client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx [hunk 8]
```diff
@@ -0,0 +1,348 @@
...
+		// in exactly the same quantities
+		if ( sorted1[ i ] !== sorted2[ i ] ) {
+			return false;
+		}
+	}
+
+	return true;
+};
+
+function useIssueAndAssignLicenses(
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx:90]: Can we break this function down into more digestible pieces? It's pretty big and hard to follow.

File: client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx (hunk 7 above)
Comment (vitozev) [reply client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx]: cc @yashwin, as I believe you worked on this, and we discussed it at some point.

File: client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx [hunk 9]
```diff
@@ -0,0 +1,348 @@
...
+
+				return page.redirect( nextStep );
+			}
+
+			// If a specific site is already selected,
+			// let's assign the licenses we just issued to it
+			const assignLicenseResponses = await assignLicensesToSite.assignLicensesToSite( issuedKeys );
+
+			const assignedLicenses = assignLicenseResponses
+				.filter( ( p ): p is PromiseFulfilledResult< APILicense > => p.status === 'fulfilled' )
```
Comment (yashwin) [inline client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx]: I think we will need both `fulfilled` and `rejected` responses to show the error message also.

File: client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx (hunk 7 above)
Comment (yashwin) [reply client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx]: Yes. We had a similar discussion where splitting the license key could be an issue, but I think I missed doing that here.

I have created a [ticket](1202619025189113-as-1205001175945323) with more info.

File: client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx (hunk 8 above)
Comment (mattgawarecki) [reply client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx:90]: I agree it's still pretty big -- this was a first attempt at splitting it up and shrinking it down without causing _too_ much disruption. 😅

Personally my approach would be to eliminate this hook entirely and rely instead on its component parts ("issue" and "assign"), moving the redirect and analytics behavior closer to the level of the rendered components that use it. _But_, I figured that was a bit of a heavy lift for one PR.

File: client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx (hunk 9 above)
Comment (mattgawarecki) [reply client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx]: I'm looking at the [corresponding code from trunk](https://github.com/Automattic/wp-calypso/blob/trunk/client/jetpack-cloud/sections/partner-portal/hooks.tsx#L423), and I don't see any error-handling code. Since I tried to preserve as much of the existing behavior as possible (to make the PR easier to review and test), I didn't address cases like this. It's something that could be added, but can you help me understand what that would look like?

Comment (yashwin) [reply client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx]: As you can see from the [corresponding code from trunk](https://github.com/Automattic/wp-calypso/blob/trunk/client/jetpack-cloud/sections/partner-portal/hooks.tsx#L423), we are not filtering the responses based on the status. We are using it [here](https://github.com/Automattic/wp-calypso/blob/trunk/client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-add-license-notification/index.tsx#L33)

Comment (mattgawarecki) [reply client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx]: Agreed that we're referencing the `status` field -- I failed to fully explain. If the promise is rejected, TypeScript asserts that the `license` field is left undefined, which means rejected Promises would never make it into the array of assigned licenses.

As for the error message -- if I'm understanding you correctly, that is -- it's dispatched as part of the mutation itself on line 82.

Comment (yashwin) [reply client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx]: Ok, could we please get rid of this [piece of code](https://github.com/Automattic/wp-calypso/blob/trunk/client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-add-license-notification/index.tsx#L33) in that case? I don't see any use of that still being there.

**Comment by mattgawarecki** [comment]
@yashwin I took some extra time after your last review comment to think a little harder about the error handling, and I think I came up with a solution. Can you take another look at this part of the hook file and tell me what you think?

https://github.com/Automattic/wp-calypso/blob/082db4d8a805359aa0c5a775049890e824da5c0c/client/jetpack-cloud/sections/partner-portal/hooks/index.tsx#L141-L172

File: client/jetpack-cloud/sections/partner-portal/hooks/index.tsx [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+		const apiRequests = keysWithProductNames.map( ( { key, name } ) =>
+			assignLicense
+				.mutateAsync( { licenseKey: key, selectedSite: selectedSiteId } )
+				.then( () => ( { key, name, status: 'fulfilled' } as ProductInfo ) )
+				.catch( () => ( { key, name, status: 'rejected' } as ProductInfo ) )
+		);
+
+		const assignLicenseStatus = {
+			selectedSite: selectedSite?.domain || '',
+			selectedProducts: await Promise.all( apiRequests ),
```
Comment (yashwin) [inline client/jetpack-cloud/sections/partner-portal/hooks/index.tsx]: I think it is better to use `Promise.allSettled()` here since we want to wait for the requests to be complete. WDYT?

File: client/jetpack-cloud/sections/partner-portal/hooks/index.tsx [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+	licenseKeys: Array< string >,
+	selectedSite: { ID: number; domain: string } | null
+): [ () => void, boolean ] {
+	const products = useProductsQuery();
+	const dispatch = useDispatch();
+	const fromDashboard = getQueryArg( window.location.href, 'source' ) === 'dashboard';
+	const assignLicense = useAssignLicenseMutation( {
+		onError: ( error: Error ) => {
+			dispatch( errorNotice( error.message, { isPersistent: true } ) );
+		},
```
Comment (yashwin) [inline client/jetpack-cloud/sections/partner-portal/hooks/index.tsx]: Now that we are handling the error in the previous way. I think we can get rid of this since this also shows error messages for all the failed API calls, which I think is redundant. 

Screenshot 2023-07-13 at 7 16 24 PM

Also, it is not clear which site it is, and there could be multiple error messages like this if multiple API calls fail.

**Review by yashwin — COMMENTED** [review]
I have tested assigning licenses, and it works as expected. Thanks for making these changes.

I have not reviewed the code changes in the PR fully, but I have left a few comments.

**Comment by mattgawarecki** [comment]
@yashwin @jkguidaven @atanas-dev I made quite a few changes and tried to incorporate all your feedback. Got time for a full re-review? 😅 

PS: I think this newer version is a lot closer to something we can safely (eventually) get rid of in favor of calling the issue and assign hooks separately. I moved a lot of UI behavior/redirects into the components that need them, so `useIssueAndAssignLicenses` is a lot less focused on those things.

**Comment by yashwin** [comment]
@mattgawarecki: This PR has grown so huge now(+ 1000+ lines), which makes it very hard to review and easy to miss any potential issues. I think it is better to split this into multiple PRs. WDYT?

**Comment by mattgawarecki** [comment]
> @mattgawarecki: This PR has grown so huge now(+ 1000+ lines), which makes it very hard to review and easy to miss any potential issues. I think it is better to split this into multiple PRs. WDYT?

I'd be happy to, except I'm not sure exactly how to break it up. These changes are all very closely tied to the same hook, and to extricate them now would be pretty difficult. If there's no other way, I can take a shot at re-writing a few new PRs to capture everything in a step-by-step approach toward the new patterns I propose, but it'll take some time and probably several more reviews.

I know this (probably) isn't the solution anybody's looking for, but one option would be to review the PR by looking at each commit in sequence. Especially toward the end of my work, I tried especially hard to make small commits to help make the changes more digestible. Does that help at all, or what is everybody thinking?

**Comment by yashwin** [comment]
> These changes are all very closely tied to the same hook, and to extricate them now would be pretty difficult

Yeah, I agree.

> I know this (probably) isn't the solution anybody's looking for, but one option would be to review the PR by looking at each commit in sequence. Especially toward the end of my work, I tried especially hard to make small commits to help make the changes more digestible.

I think this should help. I initially thought of doing this but were are 46 commits. One thing which might help here is doing an interactive rebase and trying to squash similar changes into one commit, which will reduce the no of commits. If you think it is hard too, then it is ok 👍

**Comment by mattgawarecki** [comment]
> I think this should help. I initially thought of doing this but were are 46 commits. One thing which might help here is doing an interactive rebase and trying to squash similar changes into one commit, which will reduce the no of commits. If you think it is hard too, then it is ok 👍

Doing an interactive rebase could help in some spots, but the path I took to get to where things are now was pretty winding, so squashing commits would be its own form of unraveling 😭 

In any case, I'm (re-)learning some lessons with this PR. I'll be even more merciless in the future about keeping things small 😅

**Comment by yashwin** [comment]
> Doing an interactive rebase could help in some spots, but the path I took to get to where things are now was pretty winding, so squashing commits would be its own form of unraveling

Sounds good. Could you please resolve the conflicts?

File: client/jetpack-cloud/sections/partner-portal/assign-license-form/index.tsx [hunk 12]
```diff
@@ -74,20 +98,43 @@ export default function AssignLicenseForm( {
...
+
+		dispatch( resetSite() );
+		dispatch( setPurchasedLicense( assignLicensesResult ) );
+
+		const fromDashboard = getQueryArg( window.location.href, 'source' ) === 'dashboard';
+		if ( fromDashboard ) {
+			return page.redirect( '/dashboard' );
+		}
+
+		return page.redirect( partnerPortalBasePath( '/licenses' ) );
```
Comment (yashwin) [inline client/jetpack-cloud/sections/partner-portal/assign-license-form/index.tsx:136]: Is there a reason to move all these lines here?

File: client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+			// If we know this person came from the dashboard,
+			// let's politely send them back there
+			const fromDashboard = getQueryArg( window.location.href, 'source' ) === 'dashboard';
+			if ( fromDashboard ) {
+				page.redirect( '/dashboard' );
+				return;
+			}
+
+			// Otherwise, send them to the overview of all licenses
+			page.redirect( partnerPortalBasePath( '/licenses' ) );
```
Comment (yashwin) [inline client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx:177]: Can we add a return here?

**Review by yashwin — COMMENTED** [review]
I think I tested mostly all the flows & it works as expected.

Found one issue while assigning a bundle product

There is no loading indication, and it is stuck until the requests are complete. 


Screenshot 2023-07-19 at 11 21 03 AM


I also left some comments.

I would suggest another review from someone on the team since this has a lot of changes and involves critical transaction flows.

File: client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx (hunk 13 above)
Comment (mattgawarecki) [reply client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx:177]: As the last line in a function that's typed to allow returning without a value? It's kind of implied that we're returning after the last line, but if the explicitness is important, I'll make sure it's done. Can you help me understand?

Comment (yashwin) [reply client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx:177]: Sounds good. I didn't realize that.

File: client/jetpack-cloud/sections/partner-portal/assign-license-form/index.tsx (hunk 12 above)
Comment (mattgawarecki) [reply client/jetpack-cloud/sections/partner-portal/assign-license-form/index.tsx:136]: Yeah! Generally my goal here was to separate the main concern of `useIssueAndAssignLicenses` (i.e., issuing and assigning licenses) from actions that take place visually in the browser, actions that modify the app state, or actions that record events for analytics. My desired end result was that the hook only handles the issuing and assigning of licenses, no more and no less.

I wasn't able to extract _every_ instance of these things because of how tightly coupled things were, but it's a big step in that direction.

**Comment by mattgawarecki** [comment]
> Found one issue while assigning a bundle product
>
> There is no loading indication, and it is stuck until the requests are complete.

I could be wrong, but based on my own testing, the behavior here is identical between production/staging and this PR. My approach was to throttle my network bandwidth via my browser's dev tools, and here's what I observed with throttling set to "Regular 4G/LTE" in Firefox:

- Issuing a license without a site selected: near-instantaneous.
- Selecting a site on the dashboard, clicking "Issue new license," selecting an individual product, then clicking "Issue 1 license": takes a second or so, but works without issue.
- Selecting a site on the dashboard, clicking "Issue new license," and selecting a bundle: takes long enough to notice, and there's no visible loading state.

Can you confirm for me if the behavior is different between this PR and production/staging, and what differences you observed?

---

Sidenote: I thought about going back to add a loading state for the "Select" button on bundles, but this will only make the PR even larger, so I'll address it in a separate PR. The root issue is that `LicenseBundleCard` doesn't have a concept of "loading" or "busy" the way that the `Button` component does. It'll need an extra prop added to account for this.

**Comment by yashwin** [comment]
> Can you confirm for me if the behavior is different between this PR and production/staging, and what differences you observed?

Apologies, @mattgawarecki. You are right. This behavior is found in staging/production also.

**Review by yashwin — APPROVED** [review]
LGTM. Good job on this, @mattgawarecki! 🚀

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/8068331).
